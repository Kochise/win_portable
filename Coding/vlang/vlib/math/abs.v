module math

[deprecated: 'use math.abs() instead']
pub fn fabs(x f64) f64 {
	if x > 0.0 {
		return x
	}
	return -x
}
