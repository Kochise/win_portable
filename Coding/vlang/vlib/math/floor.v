module math

// floor returns the greatest integer value less than or equal to x.
//
// special cases are:
// floor(±0) = ±0
// floor(±inf) = ±inf
// floor(nan) = nan
pub fn floor(x f64) f64 {
	if x == 0 || is_nan(x) || is_inf(x, 0) {
		return x
	}
	if x < 0 {
		mut d, fract := modf(-x)
		if fract != 0.0 {
			d = d + 1
		}
		return -d
	}
	d, _ := modf(x)
	return d
}

// ceil returns the least integer value greater than or equal to x.
//
// special cases are:
// ceil(±0) = ±0
// ceil(±inf) = ±inf
// ceil(nan) = nan
pub fn ceil(x f64) f64 {
	return -floor(-x)
}

// trunc returns the integer value of x.
//
// special cases are:
// trunc(±0) = ±0
// trunc(±inf) = ±inf
// trunc(nan) = nan
pub fn trunc(x f64) f64 {
	if x == 0 || is_nan(x) || is_inf(x, 0) {
		return x
	}
	d, _ := modf(x)
	return d
}

// round returns the nearest integer, rounding half away from zero.
//
// special cases are:
// round(±0) = ±0
// round(±inf) = ±inf
// round(nan) = nan
pub fn round(x f64) f64 {
	if x == 0 || is_nan(x) || is_inf(x, 0) {
		return x
	}
	// Largest integer <= x
	mut y := floor(x) // Fractional part
	mut r := x - y // Round up to nearest.
	if r > 0.5 {
		unsafe {
			goto rndup
		}
	}
	// Round to even
	if r == 0.5 {
		r = y - 2.0 * floor(0.5 * y)
		if r == 1.0 {
			rndup:
			y += 1.0
		}
	}
	// Else round down.
	return y
}

// round_to_even returns the nearest integer, rounding ties to even.
//
// special cases are:
// round_to_even(±0) = ±0
// round_to_even(±inf) = ±inf
// round_to_even(nan) = nan
pub fn round_to_even(x f64) f64 {
	mut bits := f64_bits(x)
	mut e_ := (bits >> shift) & mask
	if e_ >= bias {
		// round abs(x) >= 1.
		// - Large numbers without fractional components, infinity, and nan are unchanged.
		// - Add 0.499.. or 0.5 before truncating depending on whether the truncated
		// number is even or odd (respectively).
		half_minus_ulp := u64(u64(1) << (shift - 1)) - 1
		e_ -= u64(bias)
		bits += (half_minus_ulp + (bits >> (shift - e_)) & 1) >> e_
		bits &= frac_mask >> e_
		bits ^= frac_mask >> e_
	} else if e_ == bias - 1 && bits & frac_mask != 0 {
		// round 0.5 < abs(x) < 1.
		bits = bits & sign_mask | uvone // +-1
	} else {
		// round abs(x) <= 0.5 including denormals.
		bits &= sign_mask // +-0
	}
	return f64_from_bits(bits)
}
