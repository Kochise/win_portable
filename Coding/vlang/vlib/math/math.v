// Copyright (c) 2019-2021 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

// NOTE
// When adding a new function, please make sure it's in the right place.
// All functions are sorted alphabetically, separated by wrapped functions vs
// backend specific functions.
// If using System/Backend dependent functions, put them in their respective
// .c.v or .js.v or other files
// Below are functions that are not wrappers for built-in system functions, but
// native V functions. They are still sorted alphabetically
// Faster approximate sin() and cos() implemented from lolremez
pub fn aprox_sin(a f64) f64 {
	a0 := 1.91059300966915117e-31
	a1 := 1.00086760103908896
	a2 := -1.21276126894734565e-2
	a3 := -1.38078780785773762e-1
	a4 := -2.67353392911981221e-2
	a5 := 2.08026600266304389e-2
	a6 := -3.03996055049204407e-3
	a7 := 1.38235642404333740e-4
	return a0 + a * (a1 + a * (a2 + a * (a3 + a * (a4 + a * (a5 + a * (a6 + a * a7))))))
}

pub fn aprox_cos(a f64) f64 {
	a0 := 9.9995999154986614e-1
	a1 := 1.2548995793001028e-3
	a2 := -5.0648546280678015e-1
	a3 := 1.2942246466519995e-2
	a4 := 2.8668384702547972e-2
	a5 := 7.3726485210586547e-3
	a6 := -3.8510875386947414e-3
	a7 := 4.7196604604366623e-4
	a8 := -1.8776444013090451e-5
	return a0 + a * (a1 + a * (a2 + a * (a3 + a * (a4 + a * (a5 + a * (a6 + a * (a7 + a * a8)))))))
}

// copysign returns a value with the magnitude of x and the sign of y
pub fn copysign(x f64, y f64) f64 {
	return f64_from_bits((f64_bits(x) & ~sign_mask) | (f64_bits(y) & sign_mask))
}

// degrees convert from degrees to radians.
pub fn degrees(radians f64) f64 {
	return radians * (180.0 / pi)
}

// digits returns an array of the digits of n in the given base.
pub fn digits(_n int, base int) []int {
	if base < 2 {
		panic('digits: Cannot find digits of n with base $base')
	}
	mut n := _n
	mut sign := 1
	if n < 0 {
		sign = -1
		n = -n
	}
	mut res := []int{}
	for n != 0 {
		res << (n % base) * sign
		n /= base
	}
	return res
}

pub fn fabs(x f64) f64 {
	if x < 0.0 {
		return -x
	}
	return x
}

// gcd calculates greatest common (positive) divisor (or zero if a and b are both zero).
pub fn gcd(a_ i64, b_ i64) i64 {
	mut a := a_
	mut b := b_
	if a < 0 {
		a = -a
	}
	if b < 0 {
		b = -b
	}
	for b != 0 {
		a %= b
		if a == 0 {
			return b
		}
		b %= a
	}
	return a
}

// lcm calculates least common (non-negative) multiple.
pub fn lcm(a i64, b i64) i64 {
	if a == 0 {
		return a
	}
	res := a * (b / gcd(b, a))
	if res < 0 {
		return -res
	}
	return res
}

// max returns the maximum value of the two provided.
pub fn max(a f64, b f64) f64 {
	if a > b {
		return a
	}
	return b
}

// min returns the minimum value of the two provided.
pub fn min(a f64, b f64) f64 {
	if a < b {
		return a
	}
	return b
}

// radians convert from radians to degrees.
pub fn radians(degrees f64) f64 {
	return degrees * (pi / 180.0)
}
