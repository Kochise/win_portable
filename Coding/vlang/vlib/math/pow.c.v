module math

fn C.powf(x f32, y f32) f32

// powf returns base raised to the provided power. (float32)
[inline]
pub fn powf(a f32, b f32) f32 {
	return C.powf(a, b)
}
