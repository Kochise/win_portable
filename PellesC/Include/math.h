#ifndef _MATH_H
#define _MATH_H

/* math.h - standard header */

#ifndef RC_INVOKED

#if __POCC__ < 1100
#error <math.h> need POCC version 11.0 or higher
#endif /* __POCC__ < 1100 */

#pragma once

#include <crtdef.h>

#if __POCC_STDC_VERSION__ > 201710L
#define  __STDC_VERSION_MATH_H__  000000L  /* TODO */
#endif /* __POCC_STDC_VERSION__ > 201710L */

/* macros */
#define HUGE_VAL   ((double)__INFINITY__)
#define HUGE_VALF  __INFINITY__
#define HUGE_VALL  ((long double)__INFINITY__)
#define INFINITY   __INFINITY__
#define NAN        __NAN__

#define FP_INFINITE   1
#define FP_NAN        2
#define FP_NORMAL     (-1)
#define FP_SUBNORMAL  (-2)
#define FP_ZERO       0

#define FP_ILOGB0    (-0x7fffffff - 1)
#define FP_ILOGBNAN  0x7fffffff

#define MATH_ERRNO        1
#define MATH_ERREXCEPT    2
#define math_errhandling  (MATH_ERRNO|MATH_ERREXCEPT)

#define _FP_LT  1
#define _FP_EQ  2
#define _FP_GT  4

/* non-standard common math constants */
#ifdef _USE_MATH_DEFINES
#ifndef _MATH_DEFINES_DEFINED
#define _MATH_DEFINES_DEFINED
#define M_E         2.7182818284590452354   /* e */
#define M_LOG2E     1.4426950408889634074   /* log2(e) */
#define M_LOG10E    0.43429448190325182765  /* log10(e) */
#define M_LN2       0.69314718055994530942  /* log(2) */
#define M_LN10      2.30258509299404568402  /* log(10) */
#define M_PI        3.14159265358979323846  /* pi */
#define M_PI_2      1.57079632679489661923  /* pi/2 */
#define M_PI_4      0.78539816339744830962  /* pi/4 */
#define M_1_PI      0.31830988618379067154  /* 1/pi */
#define M_2_PI      0.63661977236758134308  /* 2/pi */
#define M_2_SQRTPI  1.12837916709551257390  /* 2/sqrt(pi) */
#define M_SQRT2     1.41421356237309504880  /* sqrt(2) */
#define M_SQRT1_2   0.70710678118654752440  /* 1/sqrt(2) */
#endif /* _MATH_DEFINES_DEFINED */
#endif /* _USE_MATH_DEFINES */

/* type definitions */
#if __POCC_FLTEVAL__ == 1
typedef double float_t;
typedef double double_t;
#elif __POCC_FLTEVAL__ == 2
typedef long double float_t;
typedef long double double_t;
#else /* __POCC_FLTEVAL__ <= 0 */
typedef float float_t;
typedef double double_t;
#endif

/* helper functions */
extern _CRTIMP int __cdecl __fpclass(double);
extern _CRTIMP int __cdecl __fpclassf(float);
extern _CRTIMP int __cdecl __fpclassl(long double);
extern _CRTIMP int __cdecl __fpcomp(double, double);
extern _CRTIMP int __cdecl __fpcompf(float, float);
extern _CRTIMP int __cdecl __fpcompl(long double, long double);
extern _CRTIMP int __cdecl __fpsign(double);
extern _CRTIMP int __cdecl __fpsignf(float);
extern _CRTIMP int __cdecl __fpsignl(long double);

#define isfinite(x)  (fpclassify(x) <= 0)
#define isinf(x)     (fpclassify(x) == FP_INFINITE)
#define isnan(x)     (fpclassify(x) == FP_NAN)
#define isnormal(x)  (fpclassify(x) == FP_NORMAL)

#define isgreater(x,y)  ((_FPCOMP(x,y) & _FP_GT) != 0)
#define isgreaterequal(x,y)  ((_FPCOMP(x,y) & (_FP_EQ|_FP_GT)) != 0)
#define isless(x,y)  ((_FPCOMP(x,y) & _FP_LT) != 0)
#define islessequal(x,y)  ((_FPCOMP(x,y) & (_FP_LT|_FP_EQ)) != 0)
#define islessgreater(x,y)  ((_FPCOMP(x,y) & (_FP_LT|_FP_GT)) != 0)
#define isunordered(x,y)  (_FPCOMP(x,y) == 0)

#if __POCC_STDC_VERSION__ < 201112L
#define fpclassify(x)  __generic__(x, float:__fpclassf, default:__fpclass, long double:__fpclassl)(x)
#define signbit(x)     __generic__(x, float:__fpsignf, default:__fpsign, long double:__fpsignl)(x)
#define _FPCOMP(x,y)   __generic__(x, float:__generic__(y, float:__fpcompf, default:__fpcomp, long double:__fpcompl), default:__generic__(y, default:__fpcomp, long double:__fpcompl), long double:__fpcompl)(x,y)
#else /* __POCC_STDC_VERSION__ >= 201112L */
#define fpclassify(x)  _Generic(x, float:__fpclassf, default:__fpclass, long double:__fpclassl)(x)
#define signbit(x)     _Generic(x, float:__fpsignf, default:__fpsign, long double:__fpsignl)(x)
#define _FPCOMP(x,y)   _Generic(x, float:_Generic(y, float:__fpcompf, default:__fpcomp, long double:__fpcompl), default:_Generic(y, default:__fpcomp, long double:__fpcompl), long double:__fpcompl)(x,y)
#endif /* __POCC_STDC_VERSION__ >= 201112L */

/* double declarations */
extern _CRTIMP double __cdecl acos(double);
extern _CRTIMP double __cdecl acosh(double);
extern _CRTIMP double __cdecl asin(double);
extern _CRTIMP double __cdecl asinh(double);
extern _CRTIMP double __cdecl atan(double);
extern _CRTIMP double __cdecl atan2(double, double);
extern _CRTIMP double __cdecl atanh(double);
extern _CRTIMP double __cdecl cbrt(double);
extern _CRTIMP double __cdecl ceil(double);
extern _CRTIMP double __cdecl copysign(double, double);
extern _CRTIMP double __cdecl erf(double);
extern _CRTIMP double __cdecl erfc(double);
extern _CRTIMP double __cdecl exp(double);
extern _CRTIMP double __cdecl exp2(double);
extern _CRTIMP double __cdecl expm1(double);
extern _CRTIMP double __cdecl fabs(double);
extern _CRTIMP double __cdecl fdim(double, double);
extern _CRTIMP double __cdecl floor(double);
extern _CRTIMP double __cdecl fma(double, double, double);
extern _CRTIMP double __cdecl fmax(double, double);
extern _CRTIMP double __cdecl fmin(double, double);
extern _CRTIMP double __cdecl fmod(double, double);
extern _CRTIMP double __cdecl frexp(double, int *);
extern _CRTIMP double __cdecl hypot(double, double);
extern _CRTIMP int __cdecl ilogb(double);
extern _CRTIMP double __cdecl ldexp(double, int);
extern _CRTIMP double __cdecl lgamma(double);
extern _CRTIMP long long __cdecl llrint(double);
extern _CRTIMP long long __cdecl llround(double);
extern _CRTIMP double __cdecl log1p(double);
extern _CRTIMP double __cdecl logb(double);
extern _CRTIMP long __cdecl lrint(double);
extern _CRTIMP long __cdecl lround(double);
extern _CRTIMP double __cdecl modf(double, double *);
extern _CRTIMP double __cdecl nan(const char *);
extern _CRTIMP double __cdecl nearbyint(double);
extern _CRTIMP double __cdecl nextafter(double, double);
extern _CRTIMP double __cdecl nexttoward(double, long double);
extern _CRTIMP double __cdecl pow(double, double);
extern _CRTIMP double __cdecl remainder(double, double);
extern _CRTIMP double __cdecl remquo(double, double, int *);
extern _CRTIMP double __cdecl rint(double);
extern _CRTIMP double __cdecl round(double);
extern _CRTIMP double __cdecl scalbn(double, int);
extern _CRTIMP double __cdecl scalbln(double, long);
extern _CRTIMP double __cdecl sqrt(double);
extern _CRTIMP double __cdecl tan(double);
extern _CRTIMP double __cdecl tanh(double);
extern _CRTIMP double __cdecl tgamma(double);
extern _CRTIMP double __cdecl trunc(double);

/* float declarations */
extern _CRTIMP float __cdecl acosf(float);
extern _CRTIMP float __cdecl acoshf(float);
extern _CRTIMP float __cdecl asinf(float);
extern _CRTIMP float __cdecl asinhf(float);
extern _CRTIMP float __cdecl atanf(float);
extern _CRTIMP float __cdecl atanhf(float);
extern _CRTIMP float __cdecl atan2f(float, float);
extern _CRTIMP float __cdecl cbrtf(float);
extern _CRTIMP float __cdecl ceilf(float);
extern _CRTIMP float __cdecl copysignf(float, float);
extern _CRTIMP float __cdecl erff(float);
extern _CRTIMP float __cdecl erfcf(float);
extern _CRTIMP float __cdecl expf(float);
extern _CRTIMP float __cdecl exp2f(float);
extern _CRTIMP float __cdecl expm1f(float);
extern _CRTIMP float __cdecl fabsf(float);
extern _CRTIMP float __cdecl fdimf(float, float);
extern _CRTIMP float __cdecl floorf(float);
extern _CRTIMP float __cdecl fmaf(float, float, float);
extern _CRTIMP float __cdecl fmaxf(float, float);
extern _CRTIMP float __cdecl fminf(float, float);
extern _CRTIMP float __cdecl fmodf(float, float);
extern _CRTIMP float __cdecl frexpf(float, int *);
extern _CRTIMP float __cdecl hypotf(float, float);
extern _CRTIMP int __cdecl ilogbf(float);
extern _CRTIMP float __cdecl ldexpf(float, int);
extern _CRTIMP float __cdecl lgammaf(float);
extern _CRTIMP long long __cdecl llrintf(float);
extern _CRTIMP long long __cdecl llroundf(float);
extern _CRTIMP float __cdecl log1pf(float);
extern _CRTIMP float __cdecl logbf(float);
extern _CRTIMP long __cdecl lrintf(float);
extern _CRTIMP long __cdecl lroundf(float);
extern _CRTIMP float __cdecl modff(float, float *);
extern _CRTIMP float __cdecl nanf(const char *);
extern _CRTIMP float __cdecl nearbyintf(float);
extern _CRTIMP float __cdecl nextafterf(float, float);
extern _CRTIMP float __cdecl nexttowardf(float, long double);
extern _CRTIMP float __cdecl powf(float, float);
extern _CRTIMP float __cdecl remainderf(float, float);
extern _CRTIMP float __cdecl remquof(float, float, int *);
extern _CRTIMP float __cdecl rintf(float);
extern _CRTIMP float __cdecl roundf(float);
extern _CRTIMP float __cdecl scalbnf(float, int);
extern _CRTIMP float __cdecl scalblnf(float, long);
extern _CRTIMP float __cdecl sqrtf(float);
extern _CRTIMP float __cdecl tanf(float);
extern _CRTIMP float __cdecl tanhf(float);
extern _CRTIMP float __cdecl tgammaf(float);
extern _CRTIMP float __cdecl truncf(float);

/* long double declarations */
extern _CRTIMP long double __cdecl acosl(long double);
extern _CRTIMP long double __cdecl acoshl(long double);
extern _CRTIMP long double __cdecl asinl(long double);
extern _CRTIMP long double __cdecl asinhl(long double);
extern _CRTIMP long double __cdecl atanl(long double);
extern _CRTIMP long double __cdecl atanhl(long double);
extern _CRTIMP long double __cdecl atan2l(long double, long double);
extern _CRTIMP long double __cdecl cbrtl(long double);
extern _CRTIMP long double __cdecl ceill(long double);
extern _CRTIMP long double __cdecl copysignl(long double, long double);
extern _CRTIMP long double __cdecl erfl(long double);
extern _CRTIMP long double __cdecl erfcl(long double);
extern _CRTIMP long double __cdecl expl(long double);
extern _CRTIMP long double __cdecl exp2l(long double);
extern _CRTIMP long double __cdecl expm1l(long double);
extern _CRTIMP long double __cdecl fabsl(long double);
extern _CRTIMP long double __cdecl fdiml(long double, long double);
extern _CRTIMP long double __cdecl floorl(long double);
extern _CRTIMP long double __cdecl fmal(long double, long double, long double);
extern _CRTIMP long double __cdecl fmaxl(long double, long double);
extern _CRTIMP long double __cdecl fminl(long double, long double);
extern _CRTIMP long double __cdecl fmodl(long double, long double);
extern _CRTIMP long double __cdecl frexpl(long double, int *);
extern _CRTIMP long double __cdecl hypotl(long double, long double);
extern _CRTIMP int __cdecl ilogbl(long double);
extern _CRTIMP long double __cdecl ldexpl(long double, int);
extern _CRTIMP long double __cdecl lgammal(long double);
extern _CRTIMP long long __cdecl llrintl(long double);
extern _CRTIMP long long __cdecl llroundl(long double);
extern _CRTIMP long double __cdecl log1pl(long double);
extern _CRTIMP long double __cdecl logbl(long double);
extern _CRTIMP long __cdecl lrintl(long double);
extern _CRTIMP long __cdecl lroundl(long double);
extern _CRTIMP long double __cdecl modfl(long double, long double *);
extern _CRTIMP long double __cdecl nanl(const char *);
extern _CRTIMP long double __cdecl nearbyintl(long double);
extern _CRTIMP long double __cdecl nextafterl(long double, long double);
extern _CRTIMP long double __cdecl nexttowardl(long double, long double);
extern _CRTIMP long double __cdecl powl(long double, long double);
extern _CRTIMP long double __cdecl remainderl(long double, long double);
extern _CRTIMP long double __cdecl remquol(long double, long double, int *);
extern _CRTIMP long double __cdecl rintl(long double);
extern _CRTIMP long double __cdecl roundl(long double);
extern _CRTIMP long double __cdecl scalbnl(long double, int);
extern _CRTIMP long double __cdecl scalblnl(long double, long);
extern _CRTIMP long double __cdecl sqrtl(long double);
extern _CRTIMP long double __cdecl tanl(long double);
extern _CRTIMP long double __cdecl tanhl(long double);
extern _CRTIMP long double __cdecl tgammal(long double);
extern _CRTIMP long double __cdecl truncl(long double);

extern _CRTIMP double __cdecl cos(double);
extern _CRTIMP double __cdecl cosh(double);
extern _CRTIMP double __cdecl log(double);
extern _CRTIMP double __cdecl log10(double);
extern _CRTIMP double __cdecl log2(double);
extern _CRTIMP double __cdecl sin(double);
extern _CRTIMP double __cdecl sinh(double);

extern _CRTIMP float __cdecl cosf(float);
extern _CRTIMP float __cdecl coshf(float);
extern _CRTIMP float __cdecl logf(float);
extern _CRTIMP float __cdecl log10f(float);
extern _CRTIMP float __cdecl log2f(float);
extern _CRTIMP float __cdecl sinf(float);
extern _CRTIMP float __cdecl sinhf(float);

extern _CRTIMP long double __cdecl cosl(long double);
extern _CRTIMP long double __cdecl coshl(long double);
extern _CRTIMP long double __cdecl logl(long double);
extern _CRTIMP long double __cdecl log10l(long double);
extern _CRTIMP long double __cdecl log2l(long double);
extern _CRTIMP long double __cdecl sinl(long double);
extern _CRTIMP long double __cdecl sinhl(long double);

#ifdef _MSC_EXTENSIONS
/* values for _exception type */
#define _DOMAIN  1
#define _SING  2
#define _OVERFLOW  3
#define _UNDERFLOW  4
#define _TLOSS  5
#define _PLOSS  6

/* compatibility names */
#ifdef __POCC__OLDNAMES
#define DOMAIN  _DOMAIN
#define SING  _SING
#define OVERFLOW  _OVERFLOW
#define UNDERFLOW  _UNDERFLOW
#define TLOSS  _TLOSS
#define PLOSS  _PLOSS
#endif /* __POCC__OLDNAMES */

/* passed to _matherr() when a fp exception is detected */
struct _exception {
    int type;
    const char *name;
    double arg1;
    double arg2;
    double retval;
};

extern int __cdecl _matherr(struct _exception *);
#endif /* _MSC_EXTENSIONS */

#endif /* !RC_INVOKED */

#endif /* _MATH_H */
