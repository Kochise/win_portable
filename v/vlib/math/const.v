// Copyright (c) 2019-2020 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

pub const (
	e = 2.71828182845904523536028747135266249775724709369995957496696763
	pi = 3.14159265358979323846264338327950288419716939937510582097494459
	phi = 1.61803398874989484820458683436563811772030917980576286213544862
	tau = 6.28318530717958647692528676655900576839433879875021164194988918
	sqrt2 = 1.41421356237309504880168872420969807856967187537694807317667974
	sqrt_e = 1.64872127070012814684865078781416357165377610071014801157507931
	sqrt_pi = 1.77245385090551602729816748334114518279754945612238712821380779
	sqrt_tau = 2.50662827463100050241576528481104525300698674060993831662992357
	sqrt_phi = 1.27201964951406896425242246173749149171560804184009624861664038
	ln2 = 0.693147180559945309417232121458176568075500134360255254120680009
	log2_e = 1.0 / ln2
	ln10 = 2.30258509299404568401799145468436420760110148862877297603332790
	log10_e = 1.0 / ln10
)
// Floating-point limit values
// max is the largest finite value representable by the type.
// smallest_non_zero is the smallest positive, non-zero value representable by the type.
pub const (
	max_f32 = 3.40282346638528859811704183484516925440e+38 // 2**127 * (2**24 - 1) / 2**23
	smallest_non_zero_f32 = 1.401298464324817070923729583289916131280e-45 // 1 / 2**(127 - 1 + 23)
	max_f64 = 1.797693134862315708145274237317043567981e+308 // 2**1023 * (2**53 - 1) / 2**52
	smallest_non_zero_f64 = 4.940656458412465441765687928682213723651e-324 // 1 / 2**(1023 - 1 + 52)
)
// Integer limit values
pub const (
	max_i8 = 127
	min_i8 = -128
	max_i16 = 32767
	min_i16 = -32768
	max_i32 = 2147483647
	min_i32 = -2147483648
	// -9223372036854775808 is wrong because C compilers parse litteral values
	// without sign first, and 9223372036854775808 overflows i64, hence the
	// consecutive subtraction by 1
	min_i64 = i64(-9223372036854775807 - 1)
	max_i64 = i64(9223372036854775807)
	max_u8 = 255
	max_u16 = 65535
	max_u32 = u32(4294967295)
	max_u64 = u64(18446744073709551615)
)

