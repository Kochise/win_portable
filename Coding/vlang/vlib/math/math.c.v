// Copyright (c) 2019-2021 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

#include <math.h>
$if windows {
	$if tinyc {
		#flag @VROOT/thirdparty/tcc/lib/openlibm.o
	}
}
fn C.acos(x f64) f64

fn C.asin(x f64) f64

fn C.atan(x f64) f64

fn C.atan2(y f64, x f64) f64

fn C.cbrt(x f64) f64

fn C.ceil(x f64) f64

fn C.cos(x f64) f64

fn C.cosf(x f32) f32

fn C.cosh(x f64) f64

fn C.erf(x f64) f64

fn C.erfc(x f64) f64

fn C.exp(x f64) f64

fn C.exp2(x f64) f64

fn C.fabs(x f64) f64

fn C.floor(x f64) f64

fn C.fmod(x f64, y f64) f64

fn C.hypot(x f64, y f64) f64

fn C.log(x f64) f64

fn C.log2(x f64) f64

fn C.log10(x f64) f64

fn C.lgamma(x f64) f64

fn C.pow(x f64, y f64) f64

fn C.powf(x f32, y f32) f32

fn C.round(x f64) f64

fn C.sin(x f64) f64

fn C.sinf(x f32) f32

fn C.sinh(x f64) f64

fn C.sqrt(x f64) f64

fn C.sqrtf(x f32) f32

fn C.tgamma(x f64) f64

fn C.tan(x f64) f64

fn C.tanf(x f32) f32

fn C.tanh(x f64) f64

fn C.trunc(x f64) f64

// NOTE
// When adding a new function, please make sure it's in the right place.
// All functions are sorted alphabetically.
// Returns the absolute value.
[inline]
pub fn abs(a f64) f64 {
	return C.fabs(a)
}

// acos calculates inverse cosine (arccosine).
[inline]
pub fn acos(a f64) f64 {
	return C.acos(a)
}

// asin calculates inverse sine (arcsine).
[inline]
pub fn asin(a f64) f64 {
	return C.asin(a)
}

// atan calculates inverse tangent (arctangent).
[inline]
pub fn atan(a f64) f64 {
	return C.atan(a)
}

// atan2 calculates inverse tangent with two arguments, returns the angle between the X axis and the point.
[inline]
pub fn atan2(a f64, b f64) f64 {
	return C.atan2(a, b)
}

// cbrt calculates cubic root.
[inline]
pub fn cbrt(a f64) f64 {
	return C.cbrt(a)
}

// ceil returns the nearest f64 greater or equal to the provided value.
[inline]
pub fn ceil(a f64) f64 {
	return C.ceil(a)
}

// cos calculates cosine.
[inline]
pub fn cos(a f64) f64 {
	return C.cos(a)
}

// cosf calculates cosine. (float32)
[inline]
pub fn cosf(a f32) f32 {
	return C.cosf(a)
}

// cosh calculates hyperbolic cosine.
[inline]
pub fn cosh(a f64) f64 {
	return C.cosh(a)
}

// exp calculates exponent of the number (math.pow(math.E, a)).
[inline]
pub fn exp(a f64) f64 {
	return C.exp(a)
}

// erf computes the error function value
[inline]
pub fn erf(a f64) f64 {
	return C.erf(a)
}

// erfc computes the complementary error function value
[inline]
pub fn erfc(a f64) f64 {
	return C.erfc(a)
}

// exp2 returns the base-2 exponential function of a (math.pow(2, a)).
[inline]
pub fn exp2(a f64) f64 {
	return C.exp2(a)
}

// floor returns the nearest f64 lower or equal of the provided value.
[inline]
pub fn floor(a f64) f64 {
	return C.floor(a)
}

// fmod returns the floating-point remainder of number / denom (rounded towards zero):
[inline]
pub fn fmod(a f64, b f64) f64 {
	return C.fmod(a, b)
}

// gamma computes the gamma function value
[inline]
pub fn gamma(a f64) f64 {
	return C.tgamma(a)
}

// Returns hypotenuse of a right triangle.
[inline]
pub fn hypot(a f64, b f64) f64 {
	return C.hypot(a, b)
}

// log calculates natural (base-e) logarithm of the provided value.
[inline]
pub fn log(a f64) f64 {
	return C.log(a)
}

// log2 calculates base-2 logarithm of the provided value.
[inline]
pub fn log2(a f64) f64 {
	return C.log2(a)
}

// log10 calculates the common (base-10) logarithm of the provided value.
[inline]
pub fn log10(a f64) f64 {
	return C.log10(a)
}

// log_gamma computes the log-gamma function value
[inline]
pub fn log_gamma(a f64) f64 {
	return C.lgamma(a)
}

// log_n calculates base-N logarithm of the provided value.
[inline]
pub fn log_n(a f64, b f64) f64 {
	return C.log(a) / C.log(b)
}

// pow returns base raised to the provided power.
[inline]
pub fn pow(a f64, b f64) f64 {
	return C.pow(a, b)
}

// powf returns base raised to the provided power. (float32)
[inline]
pub fn powf(a f32, b f32) f32 {
	return C.powf(a, b)
}

// round returns the integer nearest to the provided value.
[inline]
pub fn round(f f64) f64 {
	return C.round(f)
}

// sin calculates sine.
[inline]
pub fn sin(a f64) f64 {
	return C.sin(a)
}

// sinf calculates sine. (float32)
[inline]
pub fn sinf(a f32) f32 {
	return C.sinf(a)
}

// sinh calculates hyperbolic sine.
[inline]
pub fn sinh(a f64) f64 {
	return C.sinh(a)
}

// sqrt calculates square-root of the provided value.
[inline]
pub fn sqrt(a f64) f64 {
	return C.sqrt(a)
}

// sqrtf calculates square-root of the provided value. (float32)
[inline]
pub fn sqrtf(a f32) f32 {
	return C.sqrtf(a)
}

// tan calculates tangent.
[inline]
pub fn tan(a f64) f64 {
	return C.tan(a)
}

// tanf calculates tangent. (float32)
[inline]
pub fn tanf(a f32) f32 {
	return C.tanf(a)
}

// tanh calculates hyperbolic tangent.
[inline]
pub fn tanh(a f64) f64 {
	return C.tanh(a)
}

// trunc rounds a toward zero, returning the nearest integral value that is not
// larger in magnitude than a.
[inline]
pub fn trunc(a f64) f64 {
	return C.trunc(a)
}
