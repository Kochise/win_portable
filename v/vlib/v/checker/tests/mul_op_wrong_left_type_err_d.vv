import math
import math.complex as cmplx

fn main() {
	c1 := cmplx.complex(1,-2)
	c2 := c1 * 2.0
	println(c2)
}
