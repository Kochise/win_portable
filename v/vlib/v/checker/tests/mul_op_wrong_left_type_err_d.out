vlib/v/checker/tests/mul_op_wrong_left_type_err_d.vv:6:11: error: infix expr: cannot use `any_float` (right expression) as `math.complex.Complex` 
    4 | fn main() {
    5 |     c1 := cmplx.complex(1,-2)
    6 |     c2 := c1 * 2.0
      |              ^
    7 |     println(c2)
    8 | }
